magefiles/vex.go

```diff
@@ -0,0 +1,423 @@
+		}
+		findings[vex.VulnerabilityID(finding.Finding.OSV)] = finding
+	}
+	return findings, nil
+}
+
+// runGovulncheck executes the govulncheck command with the specified format
+func runGovulncheck(ctx context.Context, format string) (*bytes.Buffer, error) {
+	var buf bytes.Buffer
+	cmd := scan.Command(ctx, "-format", format, "./...")
+	cmd.Stdout = &buf
+
+	log.Info("Running govulncheck", log.String("format", format))
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("failed to start govulncheck: %w", err)
+	}
+
+	if err := cmd.Wait(); err != nil {
+		return nil, fmt.Errorf("failed to run govulncheck: %w", err)
+	}
+	return &buf, nil
+}
+
+// combineDocs merges the VEX statements from all processed tags
+func combineDocs(notAffected map[UniqueKey][]vex.Statement, affected map[UniqueKey]struct{}) []vex.Statement {
+	log.Info("Combining VEX documents")
+	statements := make(map[UniqueKey]vex.Statement)
+	for key, stmts := range notAffected {
+		for _, stmt := range stmts {
+			if _, ok := affected[key]; !ok {
+				// All versions are "not_affected" or "fixed" by the vulnerability, omitting a version in PURL
+				// => pkg:golang/github.com/aquasecurity/trivy
+				stmt.Products[0].ID = productID.String()
+				stmt.Products[0].Identifiers[vex.PURL] = productID.String()
+				statements[key] = stmt
+				break
+			}
+
+			// At least one version is "affected" by the vulnerability, so we need to include the version in PURL.
+			// => pkg:golang/github.com/aquasecurity/trivy@0.52.0
+			if s, ok := statements[key]; ok {
+				s.Products = append(s.Products, stmt.Products...)
+				statements[key] = s
+			} else {
+				statements[key] = stmt
+			}
+		}
+	}
+	return lo.Values(statements)
+}
+
+// runCommandWithTimeout executes a command with a specified timeout
+func runCommandWithTimeout(ctx context.Context, timeout time.Duration, name string, args ...string) ([]byte, error) {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, name, args...)
+	log.Info("Executing command", log.String("cmd", cmd.String()))
+
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return output, fmt.Errorf("%w, output: %s", err, string(output))
+	}
+
+	if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+		return nil, fmt.Errorf("command timed out after %v", timeout)
+	}
+
+	return output, nil
+}
+
+// updateVEX updates the final VEX document with the combined statements
+func updateVEX(output string, statements []vex.Statement) error {
+	doc, err := vex.Load(output)
+	if errors.Is(err, os.ErrNotExist) {
+		doc = &vex.VEX{}
+	} else if err != nil {
+		return err
+	}
+
+	vex.SortStatements(statements, time.Now())
+	d := &vex.VEX{
+		Metadata: vex.Metadata{
+			Context:   "https://openvex.dev/ns/v0.2.0",
+			Author:    "Aqua Security",
+			Timestamp: lo.ToPtr(time.Now()),
+			Version:   doc.Version + 1,
+			Tooling:   "https://github.com/aquasecurity/trivy/tree/main/misc/vex/vex.go",
```

Comment:
This should be updated

Reply:
Thanks. I also updated the field in the VEX file. `@id` should be re-calculated technically, but I believe it doesn't matter.
https://github.com/aquasecurity/trivy/pull/7128/commits/edfc9035cb797e55780b6f59e468f02e87101747